doc/adr/0002-pab-indexing-solution-integration.rst

```diff
@@ -0,0 +1,171 @@
+Therefore, even in this scenario, querying the plutus-chain-index with `unspentTxOutFromRef` in the above contract has a high probability of returning `Nothing`.
+
+Decision
+--------
+
+TBD
+
+Alternatives
+------------
+
+Make sure all components are in sync
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+We can make sure that all indexing solutions are syncing at the same speed.
+For example, if PAB syncs from the local node and arrives at slot 100, then it needs to wait for the chain-index to also arrive at slot 100.
+Only then can it respond to a Contract request.
+
+To be able to do that, we would need to modify the block listener of PAB and plutus-chain-index to only get new blocks once they are at the same slot.
+
+Pros:
+
+- No modification to the Contract API
+- All the indexed information is consistent with each other
+
+Cons:
+
+- As fast as the slowest indexing solution
+- Can not be feasible with external indexing solutions for which we can’t control the indexing speed (e.g. Blockfrost_ or Scrolls_)
+
+Add indexing specific functions in the Contract API
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In this scenario, we would need to split Contract API requests which interact with an external indexing solution to the ones that use the PAB.
+Currently, we have `awaitTxConfirmed` which uses the indexed information in the PAB to wait for a transaction status to change to `Confirmed`.
+On top of that, we can have `awaitTxIndexed` or `awaitTxOutIndexed` which will wait for the information to be indexed in the external indexing solution.
+
+Pros:
+
+- Limits design change on the PAB
+- More control given to the user of the Contract API
+
+Cons:
+
+- Adds an undesired complexity to the Contract API
+- We'll need to add a bunch of functions (e.g., `awaitMarconiTxConfirmed or `awaitScrollsTxConfirmed`) for each new indexing solution we want to support
+
+Ultimately, this solution is not an option given the added complexity added to the Contract API.
+
+Query functions should interact with a single source of truth
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In this scenario, we make the design decision that the Contract API should only interact with a single indexing solution.
+Thus, any blockchain information currently stored in the PAB should be moved to the indexing solution.
+Also, combining indexing solutions would need to be integrated in the single indexing solution that’s connected to the PAB.
+
+Pros:
+
+- Simplest in design to implement (other than manual work to move code)
+- No modification to the Contract API
+
+Cons:
+
+- PAB won't be able to integrate with external indexing solutions (e.g. Blockfrost_ or Scrolls_), but these external solutions would need to be integrated into the single source of truth indexing solution. Therefore, this solution reduces the PAB's flexibility.
```

Comment:
> [...] (should we show our preference here in comments?)

Good points about preference. I think it's okay to leave them here as comments. I'll add the PR in the "Notes" section of the ADR.

> It separates the semantics of the components. Indexing happens in one place, running contracts in another one.

Does this contradict what you said in the previous comment? You say here that you prefer the "single source of truth" approach, but in your previous comment you say that we should allow (eventually) users to deal with state synchronization of components. But if we have a single source of truth, then we don't need to deal with state synchronization of components.

Reply:
Yeah, at first I wrote the comment in this thread, choosing between the proposed solutions.

And later I tried to reflect about some general solution, what approach is better for us and how it fits the PAB's vision I suppose.

Reply:
IMHO whathever indexing solution PAB uses should be hidden from the user. So `unspentTxOutFromRef` (~ a PAB function) should be in charge of making sure plutus-chain-index has caught up with the rest of PAB.

Reply:
@andreabedini I'm interpreting this as solution 1 of this ADR (Make sure all components are in sync), but it might be different. You're saying that all calls to PAB functions in the Contract Monad should wait for all other components (like chain-index) to be synced and, similarly, calls to chain-index funcions should wait for all other components to be synced.

Reply:
@koslambrou isn't it "Query functions should interact with a single source of truth" ? maybe I am missing some details.
> You're saying that all calls to PAB functions in the Contract Monad should wait for all other components (like chain-index) to be synced and, similarly, calls to chain-index funcions should wait for all other components to be synced.

Yes, this is what I am thinking.

Reply:
If that's what you're thinking, then you're talking about the category "Make sure all components are in sync". With the "Query functions should interact with a single source of truth" you don't need to worry about functions not being in sync, because there is only ONE component which indexes information (all the indexing information from the PAB would be moved to that single source of truth indexing component).

Reply:
@koslambrou The way I see it is: from the POV of the contract monad, there is only one source of truth, the interpreter which is run by PAB. The interpreter giving inconsistent results is, IMHO, 100% an interpreter bug.

Reply:
Btw, I agree with the decisions in this ADR so perhaps we are just using different words to express the same concept.

Reply:
> @koslambrou The way I see it is: from the POV of the contract monad, there is only one source of truth, the interpreter which is run by PAB. The interpreter giving inconsistent results is, IMHO, 100% an interpreter bug.

Good point! I understand what you meant now.